doc/userguide/firewall/firewall-design.rst

```diff
@@ -0,0 +1,158 @@
+rule should be "hooked" into the engine. While this works well for these types of rules,
+it does lead to many edge cases that are not acceptable in an firewall ruleset. For this
+reason in the firewall rules the hook needs to be explicitly set.
+
+This is done in the protocol field of the rule. Where in threat detection a rule might look like::
+
+    alert http ... http.uri; ...
+
+In the firewall case it will be::
+
+    accept:hook http1:request_line ... http.uri; ...
+
+The application layer states / hooks are defined per protocol. Each of the hooks has it's own
+default-``drop`` policy, so a ruleset needs a ``accept`` rule for each of the states to allow
+the traffic to flow through.
+
+
+http
+^^^^
+
+For the HTTP protocol there are a number of states to hook into. These apply to HTTP 0.9, 1.0
+and 1.1. HTTP/2 uses it's own state machine.
+
+Available states:
+
+Request (``to_server``) side:
+
+* ``request_not_started``
+* ``request_line``
+* ``request_headers``
+* ``request_body``
+* ``request_trailer``
+* ``request_complete``
+
+Response (``to_client``) side:
+
+* ``response_not_started``
+* ``response_line``
+* ``response_headers``
+* ``response_body``
+* ``response_trailer``
+* ``response_complete``
+
+tls
+^^^
+
+Available states:
+
+Request (``to_server``) side:
+
+* ``client_in_progress``
+* ``client_hello_done``
+* ``client_cert_done``
+* ``client_handshake_done``
+* ``client_finished``
+
+Response (``to_client``) side:
+
+* ``server_in_progress``
+* ``server_hello``
+* ``server_cert_done``
+* ``server_hello_done``
+* ``server_handshake_done``
+* ``server_finished``
+
+Firewall pipeline
+~~~~~~~~~~~~~~~~~
+
+The firewall pipeline works in the detection engine, and is invoked after packet decoding, flow
+update, stream tracking and reassembly and app-layer parsing are all done in the context of a
+single packet.
+
+For each packet rules in the first firewall hook ``packet_filter`` are then evaluated. Assuming
+the verdict of this hook is ``accept:hook``, the next hook is evaluated: ``packet_td`` (packet
+threat detection). In this hook the IDS/IPS rules are evaluated. Rule actions here are not
+immediate, as they can still be modified by alert postprocessing like rate_filter, thresholding, etc.
+
+If the packet has been marked internally as a packet with an application layer update, then the
+next table is ``app_filter``.
+
+In ``app_filter`` the per application layer states are all evaluated at least once. At each of
+these states an ``accept:hook`` required to progress to the next state. When all available states
+have been accepted, the pipeline moves to the final table ``app_td`` (application layer threat
+detection). A ``drop`` in the ``app_filter`` table is immediate, however and ``accept`` is
+conditional on the verdict of the ``app_td`` table.
+
+In ``app_td`` the IDS/IPS rules for the application layer are evaluated. ``drop`` actions in this
+table are queued in the alert queue.
```

Comment:
For future, maybe: can it be explained why? Is it for them to generate their own alerts?